### ISSUE TITLE
[SERVE] Provide an abstraction for proxy

### ISSUE THREAD

**Issue by dioptre** [ray-project/ray#42882]
### Description

Serve is hardwired to FastAPI. 

FastAPI is 50x (not 50%) slower than other implementations like Socketify. https://github.com/cirospaciari/socketify.py

It's a matter of time before fastapi will get replaced by faster frameworks. Please abstract the proxy away so we can replace it.

### Use case

Any non-trivial sites that take a lot of load will appreciate this, saving more electricity and $ on machines, and saving the environment. The polar bears will love you.

**Comment by antoniomdk** [comment]
I think this is related to a ticket I filled recently https://github.com/ray-project/ray/issues/42392

To my mind, if Ray Serve supported any ASGI-compliant web server, that would open the door for many optimizations at the request-handling level.

**Comment by dioptre** [comment]
I think it would be nice to have support for WSGI, ASGI and unique interfaces like socketify do:
https://docs.socketify.dev/cli.html

There's reasons to not use the event loop, so not sure I'd push ASGI alone.

**Comment by dioptre** [comment]
Any updates on this? 🙏

**Comment by GeneDer** [comment]
@dioptre Nope, we do not have any near-term plan for offering this. But you are always welcome to contribute to the codebase🙂

**Comment by Superskyyy** [comment]
I'm attempting to adapt socketify.py to Ray, let's see if it works first then we can start working on the abstraction layer.

**Comment by dioptre** [comment]
Legend, let me know if you need a hand

**Comment by Superskyyy** [comment]
I tried to directly replace uvicorn with socketify's asgi class + fastapi and performance was boosted by ~10%, I guess the majority of perf gain would come when I replace the actual top framework

**Comment by dioptre** [comment]
I wonder if an intelligent cache of what queries end up where at the front would fix it and allow for more throughput. If https://github.com/ray-project/ray/issues/46693 is indeed true, that means ray serve is indeed 500x slower than socketify. That's horrifying.
